coders/qoi.c

```diff
@@ -0,0 +1,450 @@
+      px=lut[b & ~QOI_MASK_2];
+    }
+    else if ((b & QOI_MASK_2) == QOI_OP_DIFF) {
+      px.rgba.r+=((b >> 4) & 0x03) - 2;
+      px.rgba.g+=((b >> 2) & 0x03) - 2;
+      px.rgba.b+=( b       & 0x03) - 2;
+    }
+    else if ((b & QOI_MASK_2) == QOI_OP_LUMA) {
+      vg=(b & ~QOI_MASK_2) - 32;
+      if ((b=ReadBlobByte(image)) == EOF)
+        break;
+      px.rgba.r+=vg - 8 + ((b >> 4) & 0x0f);
+      px.rgba.g+=vg;
+      px.rgba.b+=vg - 8 +  (b       & 0x0f);
+    }
+    else if ((b & QOI_MASK_2) == QOI_OP_RUN) {
+      run=b & ~QOI_MASK_2;
+    }
+    lut[QOI_COLOR_HASH(px) % 64]=px;
+    do {
+      SetPixelRed(image,ScaleCharToQuantum((unsigned char)px.rgba.r),q);
+      SetPixelGreen(image,ScaleCharToQuantum((unsigned char)px.rgba.g),q);
+      SetPixelBlue(image,ScaleCharToQuantum((unsigned char)px.rgba.b),q);
+      if (channels == 4)
+        SetPixelAlpha(image,ScaleCharToQuantum((unsigned char)px.rgba.a),q);
+      q+=GetPixelChannels(image);
+      p++;
+    } while (run-- > 0);
+    status=SetImageProgress(image,LoadImageTag,p,image->rows * image->columns);
+    if (status == MagickFalse)
+      break;
+  }
+  if (EOFBlob(image) != MagickFalse)
+    ThrowFileException(exception,CorruptImageError,
+      "UnexpectedEndOfFile",image->filename);
+  if (status != MagickFalse) {
+    status=SyncAuthenticPixels(image,exception);
+    if (image->ping == MagickFalse)
+      SyncImage(image,exception);
+  }
+  (void) CloseBlob(image);
+  if (status == MagickFalse)
+    return(DestroyImageList(image));
+  return(GetFirstImageInList(image));
+}
+
+
+static MagickBooleanType WriteQOIImage(const ImageInfo *image_info,Image *image,
+  ExceptionInfo *exception)
+{
+  QuantumInfo
+    *quantum_info;
+
+  QuantumType
+    quantum_type;
+
+  MagickBooleanType
+    status;
+
+  const Quantum
+    *p;
+
+  size_t
+    channels,
+    colorspace,
+    end,
+    run,
+    idx,
+    i;
+
+  qoi_rgba_t
+    px,
+    pp,
+    lut[64];
+
+  /*
+    Open output image file.
+  */
+  assert(image_info != (const ImageInfo *) NULL);
+  assert(image_info->signature == MagickCoreSignature);
+  assert(image != (Image *) NULL);
+  assert(image->signature == MagickCoreSignature);
+  if (image->debug != MagickFalse)
+    (void) LogMagickEvent(TraceEvent,GetMagickModule(),"%s",image->filename);
+  assert(exception != (ExceptionInfo *) NULL);
+  assert(exception->signature == MagickCoreSignature);
+  status=OpenBlob(image_info,image,WriteBinaryBlobMode,exception);
+  if (status == MagickFalse)
+    return(status);
+  image->endian=MSBEndian;
+  image->depth=8;
+
+  quantum_info=AcquireQuantumInfo(image_info,image);
+  if (quantum_info == (QuantumInfo *) NULL)
+    ThrowWriterException(ImageError,"MemoryAllocationFailed");
+
+  quantum_type=GetQuantumType(image,exception);
+  if (quantum_type == RGBQuantum)
+    channels=3;
+  else if (quantum_type == RGBAQuantum)
+    channels=4;
+  else
+    ThrowWriterException(ImageError,"MemoryAllocationFailed");
```

Comment:
Is that the right exception message to throw here?

Reply:
Try `ThrowWriterException(CoderError,"ImageTypeNotSupported");`